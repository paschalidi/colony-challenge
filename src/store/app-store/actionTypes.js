export const TOGGLE_POPOVER = 'TOGGLE_POPOVER';
export const CLOSE_POPOVER = 'CLOSE_POPOVER';
export const TOGGLE_LOADING = 'TOGGLE_LOADING';

export const SELECT_AVATAR = 'SELECT_AVATAR';
export const LOAD_AVATAR = 'LOAD_AVATAR';
export const CANDIDATE_AVATAR = 'CANDIDATE_AVATAR';
export const CLEAN_CANDIDATE_AVATAR = 'CLEAN_CANDIDATE_AVATAR';

export const PRESS_ARROW_UP = 'PRESS_ARROW_UP';
export const PRESS_ARROW_DOWN = 'PRESS_ARROW_DOWN';
export const PRESS_ARROW_LEFT = 'PRESS_ARROW_LEFT';
export const PRESS_ARROW_RIGHT = 'PRESS_ARROW_RIGHT';
