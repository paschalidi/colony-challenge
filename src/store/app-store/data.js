export default function () {
  return [
    { 'src': 'https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar1.png', 'label': 'Avatar 1', 'id': 1 },
    { 'src': 'https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar2.png', 'label': 'Avatar 2', 'id': 2 },
    { 'src': 'https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar3.png', 'label': 'Avatar 3', 'id': 3 },
    { 'src': 'https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar4.png', 'label': 'Avatar 4', 'id': 4 },
    { 'src': 'https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar5.png', 'label': 'Avatar 5', 'id': 5 },
    { 'src': 'https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar6.png', 'label': 'Avatar 6', 'id': 6 }
  ];
}