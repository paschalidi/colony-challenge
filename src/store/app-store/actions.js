import * as t from './actionTypes';


export const setIndexOfActiveAvatar = (data, selectedAvatarId) => ({
  type: t.SELECT_AVATAR,
  payload: { data, selectedAvatarId }
});

export const setIndexOfLoadingAvatar = (data, selectedAvatarId) => ({
  type: t.LOAD_AVATAR,
  payload: { data, selectedAvatarId }
});

export const setIndexOfCandidateAvatar = (data, selectedAvatarId) => ({
  type: t.CANDIDATE_AVATAR,
  payload: { data, selectedAvatarId }
});

export const cleanIndexOfCandidateAvatar = () => ({
  type: t.CLEAN_CANDIDATE_AVATAR
});

//AVATAR LOADING
export const toggleLoading = () => ({
  type: t.TOGGLE_LOADING
});

//POPOVER
export const togglePopover = () => ({
  type: t.TOGGLE_POPOVER
});

export const closePopover = () => ({
  type: t.CLOSE_POPOVER
});

// KEYPRESSES
export const keypressUp = (indexOfCandidateAvatar) => {
  return {
    type: t.PRESS_ARROW_UP,
    payload: { indexOfCandidateAvatar }
  };
};

export const keypressLeft = (indexOfCandidateAvatar) => {
  return {
    type: t.PRESS_ARROW_LEFT,
    payload: { indexOfCandidateAvatar }
  };
};

export const keypressRight = (indexOfCandidateAvatar, dataLength) => {
  return {
    type: t.PRESS_ARROW_RIGHT,
    payload: { indexOfCandidateAvatar, dataLength }
  };
};

export const keypressDown = (indexOfCandidateAvatar, dataLength) => {
  return {
    type: t.PRESS_ARROW_DOWN,
    payload: { indexOfCandidateAvatar, dataLength }
  };
};

