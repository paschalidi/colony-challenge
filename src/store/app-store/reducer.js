import * as t from './actionTypes';


const INITIAL_STATE = {
  indexOfActiveAvatar: 0,
  indexOfLoadingAvatar: 0,
  indexOfCandidateAvatar: null,
  isPopoverOpen: false,
  isLoading: false
};

export const setIndexFromId = (data, selectedAvatarId) =>
  parseInt(Object.keys(data).findIndex((index) => data[index].id === selectedAvatarId))
;
export const handleKeyStrokesOnSub = (indexOfCandidateAvatar, value) => {
  if (indexOfCandidateAvatar == null)
    return 0;
  else if (indexOfCandidateAvatar !== null && indexOfCandidateAvatar - value >= 0)
    return indexOfCandidateAvatar - value;
  else
    return 0;
};
const handleKeyStrokesOnAdd = (indexOfCandidateAvatar, value, dataLength) => {
  if (indexOfCandidateAvatar == null)
    return 0;
  else if (indexOfCandidateAvatar !== null && indexOfCandidateAvatar + value < dataLength)
    return indexOfCandidateAvatar + value;
  else
    return dataLength - 1;
};
export const reducer = (state = INITIAL_STATE, { type, payload = {} }) => {
  const { indexOfCandidateAvatar, dataLength } = payload;
  switch (type) {
    case t.SELECT_AVATAR:
      return {
        ...state,
        indexOfActiveAvatar: setIndexFromId(payload.data, payload.selectedAvatarId),
        indexOfCandidateAvatar: INITIAL_STATE.indexOfCandidateAvatar
      };

    case t.LOAD_AVATAR:
      return {
        ...state,
        indexOfLoadingAvatar: setIndexFromId(payload.data, payload.selectedAvatarId)

      };

    case t.TOGGLE_LOADING:
      return {
        ...state,
        isLoading: !state.isLoading
      };

    //  POPOVER
    case t.TOGGLE_POPOVER:
      return {
        ...state,
        isPopoverOpen: !state.isPopoverOpen
      };
    case t.CLOSE_POPOVER:
      return {
        ...state,
        isPopoverOpen: INITIAL_STATE.isPopoverOpen
      };

    case t.CANDIDATE_AVATAR:
      return {
        ...state,
        indexOfCandidateAvatar: setIndexFromId(payload.data, payload.selectedAvatarId)
      };

    case t.CLEAN_CANDIDATE_AVATAR:
      return {
        ...state,
        indexOfCandidateAvatar: INITIAL_STATE.indexOfCandidateAvatar
      };


    //   KEY PRESSES
    case t.PRESS_ARROW_UP:

      return {
        ...state,
        indexOfCandidateAvatar: handleKeyStrokesOnSub(indexOfCandidateAvatar, 4)
      };
    case t.PRESS_ARROW_LEFT:
      return {
        ...state,
        indexOfCandidateAvatar: handleKeyStrokesOnSub(indexOfCandidateAvatar, 1)
      };

    case t.PRESS_ARROW_RIGHT:
      return {
        ...state,
        indexOfCandidateAvatar: handleKeyStrokesOnAdd(indexOfCandidateAvatar, 1, dataLength)
      };
    case t.PRESS_ARROW_DOWN:
      return {
        ...state,
        indexOfCandidateAvatar: handleKeyStrokesOnAdd(indexOfCandidateAvatar, 4, dataLength)
      };
  }
  return state;
};