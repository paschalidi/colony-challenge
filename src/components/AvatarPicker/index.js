/**
 *
 * AvatarPicker
 *
 */
/* eslint-disable import/first */
/* eslint-disable react/jsx-filename-extension */
/* eslint-disable react/prop-types */

import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import * as actions from '../../store/app-store/actions';
import KeyHandler, { KEYDOWN } from 'react-key-handler';
import onClickOutside from 'react-onclickoutside';


class AvatarPicker extends React.Component { // eslint-disable-line react/prefer-stateless-function
  handleClickOutside = evt => {
    evt.isTrusted && this.props.closePopover();
  };
  handleAvatarSelection = (data, selectedAvatarId) => {
    if (this.props.isLoading) return false;
    this.props.setIndexOfLoadingAvatar(data, selectedAvatarId);
    this.props.toggleLoading();
    setTimeout(() => {
      this.props.setIndexOfActiveAvatar(data, selectedAvatarId);
      this.props.toggleLoading();
      this.props.togglePopover();
    }, 1000);

  };

  render() {
    const { data, isPopoverOpen, isLoading, indexOfActiveAvatar, indexOfLoadingAvatar, indexOfCandidateAvatar } = this.props;
    return (
      <div className='container'>
        <div className='flex justify-center'>
          <div className='avatar avatar-on-top' onClick={this.props.togglePopover}>
            <img
              src={data[indexOfActiveAvatar].src}
              alt={data[indexOfActiveAvatar].label}
            />
          </div>
        </div>
        <div>
          <div className={`popover ${isPopoverOpen ? 'animation-in' : 'animation-out'}`}>
            <div className='popover-title flex justify-center'>Choose your avatar</div>
            <ul className='container flex row wrap align-start '>
              {
                Object.keys(data).map((index, key) =>
                  <li
                    key={index}
                    onClick={this.handleAvatarSelection.bind(this, data, data[index].id)}
                    onMouseEnter={() => this.props.setIndexOfCandidateAvatar(data, data[index].id)}
                    onMouseLeave={() => this.props.cleanIndexOfCandidateAvatar()}
                    className='avatar-list'
                  >
                    <div
                      className={`
                      avatar
                      ${key === indexOfActiveAvatar && 'avatar-on-list-active' }
                      ${key === indexOfCandidateAvatar && !isLoading && 'avatar-on-list-hovered avatar-on-list-hovered-background' }
                      ${key === indexOfLoadingAvatar && isLoading && 'avatar-on-list-loading' }
                    `}
                    >
                      <img src={data[index].src} alt={data[index].label} />
                    </div>
                  </li>)
              }
            </ul>
          </div>
        </div>

        <KeyHandler
          keyEventName={KEYDOWN}
          keyValue="o"
          onKeyHandle={() => this.props.togglePopover()}
        />
        {
          isPopoverOpen
          &&
          <Fragment>
            <KeyHandler
              keyEventName={KEYDOWN}
              keyValue="ArrowUp"
              onKeyHandle={() => this.props.keypressUp(indexOfCandidateAvatar)}
            />

            <KeyHandler
              keyEventName={KEYDOWN}
              keyValue="ArrowRight"
              onKeyHandle={() => this.props.keypressRight(indexOfCandidateAvatar, Object.assign(data).length)}
            />
            <KeyHandler
              keyEventName={KEYDOWN}
              keyValue="ArrowDown"
              onKeyHandle={() => this.props.keypressDown(indexOfCandidateAvatar, Object.assign(data).length)}
            />
            <KeyHandler
              keyEventName={KEYDOWN}
              keyValue="ArrowLeft"
              onKeyHandle={() => this.props.keypressLeft(indexOfCandidateAvatar)}
            />
          </Fragment>
        }
        {
          indexOfCandidateAvatar !== null
          &&
          <KeyHandler
            keyEventName={KEYDOWN}
            keyValue="Enter"
            onKeyHandle={this.handleAvatarSelection.bind(this, data, data[indexOfCandidateAvatar].id)}
          />
        }
        {/* language=CSS */}
        <style>
          {`
              .container {
                  width: 310px;
              }

              .absolute {
                  position: absolute;
                  top: 0;
                  left: 0
              }

              ul {
                  -webkit-margin-before: 0;
                  -webkit-margin-after: 0;
                  -webkit-margin-start: 0;
                  -webkit-margin-end: 0;
                  -webkit-padding-start: 0;
              }

              .avatar-list {
                  list-style-type: none;
                  padding: 7px;
              }

              .avatar,
              .avatar img {
                  width: 60px;
                  height: 60px;
                  border-radius: 50%;
                  position: relative;
              }

              .avatar-on-list-hovered::before {
                  content: '';
                  position: absolute;
                  border: 3px solid rgb(155, 160, 163) !important;
                  border-radius: 50%;
                  top: -3px;
                  bottom: -3px;
                  right: -3px;
                  left: -3px;
                  z-index: 1;
              }

              .avatar-on-list-hovered-background::after {
                  content: '';
                  position: absolute;
                  background: rgb(122, 161, 178);
                  opacity: 0.2;
                  border-radius: 50%;
                  top: 0px;
                  bottom: 0px;
                  right: 0px;
                  left: 0px;
                  z-index: 1;
              }

              .avatar-on-list-active::before {
                  content: '';
                  position: absolute;
                  border: 3px solid rgb(122, 161, 178);
                  border-radius: 50%;
                  top: -3px;
                  bottom: -3px;
                  right: -3px;
                  left: -3px;
                  z-index: 1;
              }

              .avatar-on-list-loading::before {
                  content: '';
                  position: absolute;
                  border: 3px solid rgb(122, 161, 178) !important;
                  border-bottom: 3px solid transparent !important;
                  opacity: 1 !important;
                  border-radius: 50%;
                  top: -3px;
                  bottom: -3px;
                  right: -3px;
                  left: -3px;
                  z-index: 1;
                  -webkit-animation: spin 1s linear infinite; /* Safari */
                  animation: spin 1s linear infinite;
              }

              .avatar-on-top {
                  margin: 1px 0;
              }

              .avatar-on-top:hover::before {
                  content: '';
                  position: absolute;
                  border: 1px solid rgb(155, 160, 163) !important;
                  border-radius: 50%;
                  top: -1px;
                  bottom: -1px;
                  right: -1px;
                  left: -1px;
                  z-index: 1;
              }

              .popover {
                  border-radius: 2px;
                  background: rgb(44, 48, 51);
                  box-shadow: 2px 2px 10px 2px rgb(102, 102, 102);
                  padding: 7px 7px;
                  height: auto;
                  margin-top: 10px;
              }

              .popover::after {
                  border-bottom: 8px solid rgb(44, 48, 51);
                  border-left: 8px solid transparent;
                  border-right: 8px solid transparent;
                  width: 0;
                  height: 0;

                  content: "";
                  display: block;
                  position: absolute;
                  bottom: 100%;
                  left: calc(50% - 8px);
              }

              .popover-title {
                  font-family: 'Source Sans Pro', sans-serif;
                  color: rgb(249, 249, 249);
                  font-size: 16px;
                  margin: 10px 0;

              }

              .animation-in {
                  opacity: 0;
                  animation-name: animationIn;
                  animation-duration: 200ms;
                  animation-timing-function: linear;
                  animation-fill-mode: forwards;
              }

              .animation-out {
                  opacity: 0;
                  animation-name: animationOut;
                  animation-duration: 100ms;
                  animation-timing-function: linear;
                  animation-fill-mode: forwards;
              }

              @keyframes animationIn {
                  0% {
                      opacity: 0;
                      transform: scale(0.3) translate3d(0, 0, 0);
                  }
                  50% {
                      opacity: 0.9;
                      transform: scale(1.02);
                  }
                  95% {
                      opacity: 1;
                      transform: scale(0.95);
                  }
                  100% {
                      opacity: 1;
                      transform: scale(1) translate3d(0, 0, 0);
                  }
              }

              @keyframes animationOut {
                  0% {
                      opacity: 1;
                      transform: scale(1) translate3d(0, 0, 0);

                  }
                  100% {
                      opacity: 0;
                      transform: scale(0) translate3d(0, 0, 0);

                  }
              }

              /* Safari */
              @-webkit-keyframes spin {
                  0% {
                      -webkit-transform: rotate(0deg);
                  }
                  100% {
                      -webkit-transform: rotate(360deg);
                  }
              }

              @keyframes spin {
                  0% {
                      transform: rotate(0deg);
                  }
                  100% {
                      transform: rotate(360deg);
                  }
              }
          `}
        </style>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    data: state.data,
    ...state.reducer
  };
}

export default connect(mapStateToProps, actions)(onClickOutside(AvatarPicker));