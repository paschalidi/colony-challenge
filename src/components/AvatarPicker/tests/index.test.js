import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import AvatarPicker from '../index';

import Store from '../../../store/configure';
import {
  reducer,
  handleKeyStrokesOnSub,
  handleKeyStrokesOnAdd
} from '../../../store/app-store/index';
import * as t from '../../../store/app-store/actionTypes';


const mockingData = [
  { 'label': 'Avatar 5', 'id': 10 },
  { 'label': 'Avatar 6', 'id': 11 },
  { 'label': 'Avatar 1', 'id': 1 },
  { 'label': 'Avatar 2', 'id': 2 },
  { 'label': 'Avatar 3', 'id': 3 },
  { 'label': 'Avatar 4', 'id': 4 },
  { 'label': 'Avatar 5', 'id': 5 },
  { 'label': 'Avatar 6', 'id': 6 },
  { 'label': 'Avatar 2', 'id': 7 },
  { 'label': 'Avatar 3', 'id': 8 },
  { 'label': 'Avatar 4', 'id': 9 }
];

const StoreInstance = Store();

test('AvatarPicker renders', () => {
  const component = renderer.create(
    <Provider store={StoreInstance}>
      <AvatarPicker />
    </Provider>
  );
  let tree;

  tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

//  REDUCER
describe('app reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual({
        indexOfActiveAvatar: 0,
        indexOfLoadingAvatar: 0,
        indexOfCandidateAvatar: null,
        isPopoverOpen: false,
        isLoading: false
      }
    );
  });

  it('should handle SELECT_AVATAR', () => {
    expect(
      reducer({}, {
        type: t.SELECT_AVATAR,
        payload: { data: mockingData, selectedAvatarId: 10 }
      })
    ).toEqual({
        indexOfActiveAvatar: 0,
        indexOfCandidateAvatar: null
      }
    );
  });

  it('should handle LOAD_AVATAR', () => {
    expect(
      reducer({}, {
        type: t.LOAD_AVATAR,
        payload: { data: mockingData, selectedAvatarId: 11 }
      })
    ).toEqual({
        indexOfLoadingAvatar: 1
      }
    );
  });

  it('should handle CLOSE_POPOVER', () => {
    expect(
      reducer({}, {
        type: t.CLOSE_POPOVER
      })
    ).toEqual({
        isPopoverOpen: false
      }
    );
  });
  it('should handle CANDIDATE_AVATAR', () => {
    expect(
      reducer({}, {
        type: t.CANDIDATE_AVATAR,
        payload: { data: mockingData, selectedAvatarId: 5 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 6
      }
    );
  });
  it('should handle CLEAN_CANDIDATE_AVATAR', () => {
    expect(
      reducer({}, {
        type: t.CLEAN_CANDIDATE_AVATAR
      })
    ).toEqual({
        indexOfCandidateAvatar: null
      }
    );
  });
  it('should handle PRESS_ARROW_UP', () => {
    expect(
      reducer({}, {
        type: t.PRESS_ARROW_UP,
        payload: { indexOfCandidateAvatar: 5 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 1
      }
    );
  });
  it('should handle PRESS_ARROW_UP', () => {
    expect(
      reducer({}, {
        type: t.PRESS_ARROW_UP,
        payload: { indexOfCandidateAvatar: 3 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 0
      }
    );
  });
  it('should handle PRESS_ARROW_LEFT', () => {
    expect(
      reducer({}, {
        type: t.PRESS_ARROW_LEFT,
        payload: { indexOfCandidateAvatar: 5 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 4
      }
    );
  });

  it('should handle PRESS_ARROW_LEFT', () => {
    expect(
      reducer({}, {
        type: t.PRESS_ARROW_LEFT,
        payload: { indexOfCandidateAvatar: 0 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 0
      }
    );
  });
  it('should handle PRESS_ARROW_RIGHT 10', () => {
    expect(
      reducer({}, {
        type: t.PRESS_ARROW_RIGHT,
        payload: { indexOfCandidateAvatar: 10, dataLength: 11 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 10
      }
    );
  });
  it('should handle PRESS_ARROW_RIGHT 1', () => {
    expect(
      reducer({}, {
        type: t.PRESS_ARROW_RIGHT,
        payload: { indexOfCandidateAvatar: 1, dataLength: 11 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 2
      }
    );
  });
  it('should handle PRESS_ARROW_DOWN', () => {
    expect(
      reducer({}, {
        type: t.PRESS_ARROW_DOWN,
        payload: { indexOfCandidateAvatar: 2, dataLength: 11 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 6
      }
    );
  });
  it('should handle PRESS_ARROW_DOWN', () => {
    expect(
      reducer({}, {
        type: t.PRESS_ARROW_DOWN,
        payload: { indexOfCandidateAvatar: 19, dataLength: 11 }
      })
    ).toEqual({
        indexOfCandidateAvatar: 10
      }
    );
  });
});


