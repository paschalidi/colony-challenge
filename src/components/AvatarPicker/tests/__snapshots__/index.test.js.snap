// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarPicker renders 1`] = `
<div
  className="container"
>
  <div
    className="flex justify-center"
  >
    <div
      className="avatar avatar-on-top"
      onClick={[Function]}
    >
      <img
        alt="Avatar 1"
        src="https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar1.png"
      />
    </div>
  </div>
  <div>
    <div
      className="popover animation-out"
    >
      <div
        className="popover-title flex justify-center"
      >
        Choose your avatar
      </div>
      <ul
        className="container flex row wrap align-start "
      >
        <li
          className="avatar-list"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="
                                avatar
                                avatar-on-list-active
                                false
                                false
                              "
          >
            <img
              alt="Avatar 1"
              src="https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar1.png"
            />
          </div>
        </li>
        <li
          className="avatar-list"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="
                                avatar
                                false
                                false
                                false
                              "
          >
            <img
              alt="Avatar 2"
              src="https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar2.png"
            />
          </div>
        </li>
        <li
          className="avatar-list"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="
                                avatar
                                false
                                false
                                false
                              "
          >
            <img
              alt="Avatar 3"
              src="https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar3.png"
            />
          </div>
        </li>
        <li
          className="avatar-list"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="
                                avatar
                                false
                                false
                                false
                              "
          >
            <img
              alt="Avatar 4"
              src="https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar4.png"
            />
          </div>
        </li>
        <li
          className="avatar-list"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="
                                avatar
                                false
                                false
                                false
                              "
          >
            <img
              alt="Avatar 5"
              src="https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar5.png"
            />
          </div>
        </li>
        <li
          className="avatar-list"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="
                                avatar
                                false
                                false
                                false
                              "
          >
            <img
              alt="Avatar 6"
              src="https://raw.githubusercontent.com/JoinColony/coding-challenge/master/avatar6.png"
            />
          </div>
        </li>
      </ul>
    </div>
  </div>
  <style>
    
                  .container {
                      width: 310px;
                  }
    
                  .absolute {
                      position: absolute;
                      top: 0;
                      left: 0
                  }
    
                  ul {
                      -webkit-margin-before: 0;
                      -webkit-margin-after: 0;
                      -webkit-margin-start: 0;
                      -webkit-margin-end: 0;
                      -webkit-padding-start: 0;
                  }
    
                  .avatar-list {
                      list-style-type: none;
                      padding: 7px;
                  }
    
                  .avatar,
                  .avatar img {
                      width: 60px;
                      height: 60px;
                      border-radius: 50%;
                      position: relative;
                  }
    
                  .avatar-on-list-hovered::before {
                      content: '';
                      position: absolute;
                      border: 3px solid rgb(155, 160, 163) !important;
                      border-radius: 50%;
                      top: -3px;
                      bottom: -3px;
                      right: -3px;
                      left: -3px;
                      z-index: 1;
                  }
    
                  .avatar-on-list-hovered-background::after {
                      content: '';
                      position: absolute;
                      background: rgb(122, 161, 178);
                      opacity: 0.2;
                      border-radius: 50%;
                      top: 0px;
                      bottom: 0px;
                      right: 0px;
                      left: 0px;
                      z-index: 1;
                  }
    
                  .avatar-on-list-active::before {
                      content: '';
                      position: absolute;
                      border: 3px solid rgb(122, 161, 178);
                      border-radius: 50%;
                      top: -3px;
                      bottom: -3px;
                      right: -3px;
                      left: -3px;
                      z-index: 1;
                  }
    
                  .avatar-on-list-loading::before {
                      content: '';
                      position: absolute;
                      border: 3px solid rgb(122, 161, 178) !important;
                      border-bottom: 3px solid transparent !important;
                      opacity: 1 !important;
                      border-radius: 50%;
                      top: -3px;
                      bottom: -3px;
                      right: -3px;
                      left: -3px;
                      z-index: 1;
                      -webkit-animation: spin 1s linear infinite; /* Safari */
                      animation: spin 1s linear infinite;
                  }
    
                  .avatar-on-top {
                      margin: 1px 0;
                  }
    
                  .avatar-on-top:hover::before {
                      content: '';
                      position: absolute;
                      border: 1px solid rgb(155, 160, 163) !important;
                      border-radius: 50%;
                      top: -1px;
                      bottom: -1px;
                      right: -1px;
                      left: -1px;
                      z-index: 1;
                  }
    
                  .popover {
                      border-radius: 2px;
                      background: rgb(44, 48, 51);
                      box-shadow: 2px 2px 10px 2px rgb(102, 102, 102);
                      padding: 7px 7px;
                      height: auto;
                      margin-top: 10px;
                  }
    
                  .popover::after {
                      border-bottom: 8px solid rgb(44, 48, 51);
                      border-left: 8px solid transparent;
                      border-right: 8px solid transparent;
                      width: 0;
                      height: 0;
    
                      content: "";
                      display: block;
                      position: absolute;
                      bottom: 100%;
                      left: calc(50% - 8px);
                  }
    
                  .popover-title {
                      font-family: 'Source Sans Pro', sans-serif;
                      color: rgb(249, 249, 249);
                      font-size: 16px;
                      margin: 10px 0;
    
                  }
    
                  .animation-in {
                      opacity: 0;
                      animation-name: animationIn;
                      animation-duration: 200ms;
                      animation-timing-function: linear;
                      animation-fill-mode: forwards;
                  }
    
                  .animation-out {
                      opacity: 0;
                      animation-name: animationOut;
                      animation-duration: 100ms;
                      animation-timing-function: linear;
                      animation-fill-mode: forwards;
                  }
    
                  @keyframes animationIn {
                      0% {
                          opacity: 0;
                          transform: scale(0.3) translate3d(0, 0, 0);
                      }
                      50% {
                          opacity: 0.9;
                          transform: scale(1.02);
                      }
                      95% {
                          opacity: 1;
                          transform: scale(0.95);
                      }
                      100% {
                          opacity: 1;
                          transform: scale(1) translate3d(0, 0, 0);
                      }
                  }
    
                  @keyframes animationOut {
                      0% {
                          opacity: 1;
                          transform: scale(1) translate3d(0, 0, 0);
    
                      }
                      100% {
                          opacity: 0;
                          transform: scale(0) translate3d(0, 0, 0);
    
                      }
                  }
    
                  /* Safari */
                  @-webkit-keyframes spin {
                      0% {
                          -webkit-transform: rotate(0deg);
                      }
                      100% {
                          -webkit-transform: rotate(360deg);
                      }
                  }
    
                  @keyframes spin {
                      0% {
                          transform: rotate(0deg);
                      }
                      100% {
                          transform: rotate(360deg);
                      }
                  }
              
  </style>
</div>
`;
